cpp/src/arrow/compute/kernels/vector_swizzle_test.cc

```diff
@@ -0,0 +1,1065 @@
+    auto indices = "[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]";
+    int64_t max_index = 9;
+    auto expected = R"(["j", null, "h", null, "f", null, "d", null, "b", null])";
+    this->TestScatter(values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Indices with nulls");
+    auto values = R"(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"])";
+    auto indices = "[9, null, 7, null, 5, null, 3, null, 1, null]";
+    int64_t max_index = 9;
+    auto expected = R"([null, "i", null, "g", null, "e", null, "c", null, "a"])";
+    this->TestScatter(values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Indices with negatives");
+    auto values = R"(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"])";
+    auto indices = "[9, -1, 7, -2, 5, -3, 3, -4, 1, -5]";
+    int64_t max_index = 9;
+    auto expected = R"([null, "i", null, "g", null, "e", null, "c", null, "a"])";
+    this->TestScatterSignedIndices(values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Output less than input");
+    auto values = R"(["b", "a"])";
+    auto indices = "[1, 0]";
+    int64_t max_index = 0;
+    auto expected = R"(["a"])";
+    this->TestScatter(values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Output greater than input");
+    auto values = R"(["a", "a", "a", "b", "b", "b"])";
+    auto indices = "[0, 3, 6, 1, 4, 7]";
+    int64_t max_index = 8;
+    auto expected = R"(["a", "b", null, "a", "b", null, "a", "b", null])";
+    this->TestScatter(values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Values all null");
+    auto values = "[null, null]";
+    auto indices = "[0, 1]";
+    int64_t max_index = 1;
+    auto expected = "[null, null]";
+    this->TestScatter(values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Indices all null");
+    auto values = R"(["a", "b"])";
+    auto indices = "[null, null]";
+    int64_t max_index = 1;
+    auto expected = "[null, null]";
+    this->TestScatter(values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Output all null");
+    auto values = R"(["a", "b"])";
+    auto indices = "[2, 3]";
+    int64_t max_index = 1;
+    auto expected = "[null, null]";
+    this->TestScatter(values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Empty input output null");
+    auto values = "[]";
+    auto indices = "[]";
+    int64_t max_index = 1;
+    auto expected = "[null, null]";
+    this->TestScatter(values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Indices duplicated indices");
+    auto values = R"(["a", "b", null, null])";
+    auto indices = "[0, 1, 0, 1]";
+    int64_t max_index = 3;
+    auto expected = "[null, null, null, null]";
+    this->TestScatter(values, indices, max_index, expected);
+  }
+}
+
+// ----------------------------------------------------------------------
+// Test Scatter using a hypothetical if-else special form.
```

Comment:
What is a "special form"? Is this a specific technical term (can you point to a definition)?

Reply:
This is the original rationale of special form: https://github.com/apache/arrow/issues/41094#issuecomment-2087716483

And a copy-paste of the comment from my [previous (immature) attempt](https://github.com/apache/arrow/pull/42106) to implement special form:
```
/// The concept "special form" is borrowed from Lisp
/// (https://courses.cs.northwestern.edu/325/readings/special-forms.html). Velox also uses
/// the same term. A special form behaves like a function call except that it has special
/// evaluation rules, mostly for arguments.
/// For example, the `if_else(cond, expr1, expr2)` special form first evaluates the
/// argument `cond` and obtains a boolean array:
///   [true, false, true, false]
/// then the argument `expr1` should ONLY be evaluated for row:
///   [0, 2]
/// and the argument `expr2` should ONLY be evaluated for row:
///   [1, 3]
/// Consider, if `expr1`/`expr2` has some observable side-effects (e.g., division by zero
/// error) on row [1, 3]/[0, 2], these side-effects would be undesirably observed if
/// evaluated using a regular function call, which always evaluates all its arguments
/// eagerly.
/// Other special forms include `case_when`, `and`, and `or`, etc.
/// In a vectorized execution engine, a special form normally takes advantage of
/// "selection vector" to mask rows of arguments to be evaluated.
```

Reply:
Ok, so it's conditional evaluation? And gather-evaluate-scatter is an alternative to using selection vectors?

Reply:
Yes. This will be an incremental approach to allow all the kernels gradually evolve to be aware of "selection vector".

Quoting @felipecrv :
> It's unrealistic to expect we will specialize all kernel functions to handle the optional selection vector parameter, but we should at least push them all the way down to Function::Execute. If a Function doesn't handle selection vectors, then a call is dispatched for all the values (as it is today) and code similar to Take is used to gather the values that the selection vector selects. The more functions become aware of selection vectors, the less we have to rely on this slow code.